### PR TITLE
Fingerprint Authentication

### PR DESCRIPTION
**Pull request by Achintya999**
Added Fingerprint Authentication in the App.
#1380

### CODE REVIEW

File: app/src/main/java/org/mozilla/focus/activity/FingerprintActivity.java [hunk 1]
```diff
@@ -0,0 +1,145 @@
...
+    private static final String KEY_NAME = "minor";
+    private Cipher cipher;
+    private TextView textView;
+
+    @Override
+    protected void onCreate(Bundle savedInstanceState) {
+        super.onCreate(savedInstanceState);
+        setContentView(R.layout.activity_fingerprint);
+
+        KeyguardManager keyguardManager = (KeyguardManager)getSystemService(KEYGUARD_SERVICE);
```
Comment (sputnikci) [inline app/src/main/java/org/mozilla/focus/activity/FingerprintActivity.java]: \[Checkstyle\] ERROR: 'typecast' is not followed by whitespace\.

File: app/src/main/java/org/mozilla/focus/activity/FingerprintActivity.java [hunk 2]
```diff
@@ -0,0 +1,145 @@
...
+        FingerprintManager fingerprintManager = (FingerprintManager)getSystemService(FINGERPRINT_SERVICE);
```
Comment (sputnikci) [inline app/src/main/java/org/mozilla/focus/activity/FingerprintActivity.java]: \[Checkstyle\] ERROR: 'typecast' is not followed by whitespace\.

File: app/src/main/java/org/mozilla/focus/activity/FingerprintActivity.java [hunk 3]
```diff
@@ -0,0 +1,145 @@
...
+
+        textView = (TextView)findViewById(R.id.errorText);
```
Comment (sputnikci) [inline app/src/main/java/org/mozilla/focus/activity/FingerprintActivity.java]: \[Checkstyle\] ERROR: 'typecast' is not followed by whitespace\.

File: app/src/main/java/org/mozilla/focus/activity/FingerprintActivity.java [hunk 4]
```diff
@@ -0,0 +1,145 @@
...
+
+        if(!fingerprintManager.isHardwareDetected()){
```
Comment (sputnikci) [inline app/src/main/java/org/mozilla/focus/activity/FingerprintActivity.java]: \[Checkstyle\] ERROR: 'if' is not followed by whitespace\.

Comment (sputnikci) [inline app/src/main/java/org/mozilla/focus/activity/FingerprintActivity.java]: \[Checkstyle\] ERROR: '\{' is not preceded with whitespace\.

File: app/src/main/java/org/mozilla/focus/activity/FingerprintActivity.java [hunk 5]
```diff
@@ -0,0 +1,145 @@
...
+
+            textView.setText("Your Device Does Not Have A FingerPrint Sensor");
+//            Intent intent = new Intent(getApplicationContext(), MainActivity.class);
+
+        }else {
```
Comment (sputnikci) [inline app/src/main/java/org/mozilla/focus/activity/FingerprintActivity.java]: \[Checkstyle\] ERROR: 'else' is not preceded with whitespace\.

Comment (sputnikci) [inline app/src/main/java/org/mozilla/focus/activity/FingerprintActivity.java]: \[Checkstyle\] ERROR: '\}' is not followed by whitespace\.

File: app/src/main/java/org/mozilla/focus/activity/FingerprintActivity.java [hunk 6]
```diff
@@ -0,0 +1,145 @@
...
+
+            if(ActivityCompat.checkSelfPermission(this, Manifest.permission.USE_FINGERPRINT) != PackageManager.PERMISSION_GRANTED){
```
Comment (sputnikci) [inline app/src/main/java/org/mozilla/focus/activity/FingerprintActivity.java]: \[Checkstyle\] ERROR: 'if' is not followed by whitespace\.

Comment (sputnikci) [inline app/src/main/java/org/mozilla/focus/activity/FingerprintActivity.java]: \[Checkstyle\] ERROR: '\{' is not preceded with whitespace\.

File: app/src/main/java/org/mozilla/focus/activity/FingerprintActivity.java [hunk 7]
```diff
@@ -0,0 +1,145 @@
...
+
+                textView.setText("FingerPrint Authentication Permission Not Enabled");
+            }else {
```
Comment (sputnikci) [inline app/src/main/java/org/mozilla/focus/activity/FingerprintActivity.java]: \[Checkstyle\] ERROR: 'else' is not preceded with whitespace\.

Comment (sputnikci) [inline app/src/main/java/org/mozilla/focus/activity/FingerprintActivity.java]: \[Checkstyle\] ERROR: '\}' is not followed by whitespace\.

File: app/src/main/java/org/mozilla/focus/activity/FingerprintActivity.java [hunk 8]
```diff
@@ -0,0 +1,145 @@
...
+
+                if(!fingerprintManager.hasEnrolledFingerprints()){
```
Comment (sputnikci) [inline app/src/main/java/org/mozilla/focus/activity/FingerprintActivity.java]: \[Checkstyle\] ERROR: 'if' is not followed by whitespace\.

Comment (sputnikci) [inline app/src/main/java/org/mozilla/focus/activity/FingerprintActivity.java]: \[Checkstyle\] ERROR: '\{' is not preceded with whitespace\.

File: app/src/main/java/org/mozilla/focus/activity/FingerprintActivity.java [hunk 9]
```diff
@@ -0,0 +1,145 @@
...
+
+                    textView.setText("Register Atleast One FingerPrint In Settings");
+                }else {
```
Comment (sputnikci) [inline app/src/main/java/org/mozilla/focus/activity/FingerprintActivity.java]: \[Checkstyle\] ERROR: 'else' is not preceded with whitespace\.

Comment (sputnikci) [inline app/src/main/java/org/mozilla/focus/activity/FingerprintActivity.java]: \[Checkstyle\] ERROR: '\}' is not followed by whitespace\.

File: app/src/main/java/org/mozilla/focus/activity/FingerprintActivity.java [hunk 10]
```diff
@@ -0,0 +1,145 @@
...
+
+                    if(!keyguardManager.isKeyguardSecure()){
```
Comment (sputnikci) [inline app/src/main/java/org/mozilla/focus/activity/FingerprintActivity.java]: \[Checkstyle\] ERROR: 'if' is not followed by whitespace\.

Comment (sputnikci) [inline app/src/main/java/org/mozilla/focus/activity/FingerprintActivity.java]: \[Checkstyle\] ERROR: '\{' is not preceded with whitespace\.

File: app/src/main/java/org/mozilla/focus/activity/FingerprintActivity.java [hunk 11]
```diff
@@ -0,0 +1,145 @@
...
+                        textView.setText("Lock Screen Security Not Enabled In Settings");
+                    }else {
```
Comment (sputnikci) [inline app/src/main/java/org/mozilla/focus/activity/FingerprintActivity.java]: \[Checkstyle\] ERROR: 'else' is not preceded with whitespace\.

Comment (sputnikci) [inline app/src/main/java/org/mozilla/focus/activity/FingerprintActivity.java]: \[Checkstyle\] ERROR: '\}' is not followed by whitespace\.

File: app/src/main/java/org/mozilla/focus/activity/FingerprintActivity.java [hunk 12]
```diff
@@ -0,0 +1,145 @@
...
+                        generateKey();
+
+                        if (cipherInit()){
```
Comment (sputnikci) [inline app/src/main/java/org/mozilla/focus/activity/FingerprintActivity.java]: \[Checkstyle\] ERROR: '\{' is not preceded with whitespace\.

File: app/src/main/java/org/mozilla/focus/activity/FingerprintActivity.java [hunk 13]
```diff
@@ -0,0 +1,145 @@
...
+                    }
+                }
+            }
+
+
+        }
+    }
+
+    @TargetApi(Build.VERSION_CODES.M)
+    protected void generateKey(){
```
Comment (sputnikci) [inline app/src/main/java/org/mozilla/focus/activity/FingerprintActivity.java]: \[Checkstyle\] ERROR: '\{' is not preceded with whitespace\.

File: app/src/main/java/org/mozilla/focus/activity/FingerprintActivity.java [hunk 14]
```diff
@@ -0,0 +1,145 @@
...
+
+        try{
```
Comment (sputnikci) [inline app/src/main/java/org/mozilla/focus/activity/FingerprintActivity.java]: \[Checkstyle\] ERROR: 'try' is not followed by whitespace\.

Comment (sputnikci) [inline app/src/main/java/org/mozilla/focus/activity/FingerprintActivity.java]: \[Checkstyle\] ERROR: '\{' is not preceded with whitespace\.

File: app/src/main/java/org/mozilla/focus/activity/FingerprintActivity.java [hunk 15]
```diff
@@ -0,0 +1,145 @@
...
+            keyStore = KeyStore.getInstance("AndroidKeyStore");
+        }catch (Exception e){
```
Comment (sputnikci) [inline app/src/main/java/org/mozilla/focus/activity/FingerprintActivity.java]: \[Checkstyle\] ERROR: 'catch' is not preceded with whitespace\.

Comment (sputnikci) [inline app/src/main/java/org/mozilla/focus/activity/FingerprintActivity.java]: \[Checkstyle\] ERROR: '\}' is not followed by whitespace\.

Comment (sputnikci) [inline app/src/main/java/org/mozilla/focus/activity/FingerprintActivity.java]: \[Checkstyle\] ERROR: '\{' is not preceded with whitespace\.

File: app/src/main/java/org/mozilla/focus/activity/FingerprintActivity.java [hunk 16]
```diff
@@ -0,0 +1,145 @@
...
+            e.printStackTrace();
+        }
+
+        KeyGenerator keyGenerator;
+        try{
```
Comment (sputnikci) [inline app/src/main/java/org/mozilla/focus/activity/FingerprintActivity.java]: \[Checkstyle\] ERROR: 'try' is not followed by whitespace\.

Comment (sputnikci) [inline app/src/main/java/org/mozilla/focus/activity/FingerprintActivity.java]: \[Checkstyle\] ERROR: '\{' is not preceded with whitespace\.

File: app/src/main/java/org/mozilla/focus/activity/FingerprintActivity.java [hunk 17]
```diff
@@ -0,0 +1,145 @@
...
+            keyGenerator = KeyGenerator.getInstance(KeyProperties.KEY_ALGORITHM_AES, "AndroidKeyStore");
+        }catch (NoSuchAlgorithmException | NoSuchProviderException e){
```
Comment (sputnikci) [inline app/src/main/java/org/mozilla/focus/activity/FingerprintActivity.java]: \[Checkstyle\] ERROR: 'catch' is not preceded with whitespace\.

Comment (sputnikci) [inline app/src/main/java/org/mozilla/focus/activity/FingerprintActivity.java]: \[Checkstyle\] ERROR: '\}' is not followed by whitespace\.

Comment (sputnikci) [inline app/src/main/java/org/mozilla/focus/activity/FingerprintActivity.java]: \[Checkstyle\] ERROR: '\{' is not preceded with whitespace\.

File: app/src/main/java/org/mozilla/focus/activity/FingerprintActivity.java [hunk 18]
```diff
@@ -0,0 +1,145 @@
...
+
+            throw new RuntimeException("Failed To Get Key Generator Instance", e);
+        }
+
+        try{
```
Comment (sputnikci) [inline app/src/main/java/org/mozilla/focus/activity/FingerprintActivity.java]: \[Checkstyle\] ERROR: 'try' is not followed by whitespace\.

Comment (sputnikci) [inline app/src/main/java/org/mozilla/focus/activity/FingerprintActivity.java]: \[Checkstyle\] ERROR: '\{' is not preceded with whitespace\.

File: app/src/main/java/org/mozilla/focus/activity/FingerprintActivity.java [hunk 19]
```diff
@@ -0,0 +1,145 @@
...
+                    KeyGenParameterSpec.Builder(KEY_NAME,
+                    KeyProperties.PURPOSE_ENCRYPT |
+                            KeyProperties.PURPOSE_DECRYPT)
+                    .setBlockModes(KeyProperties.BLOCK_MODE_CBC)
+                    .setUserAuthenticationRequired(true)
+                    .setEncryptionPaddings(
+                            KeyProperties.ENCRYPTION_PADDING_PKCS7)
+                    .build());
+            keyGenerator.generateKey();
+        }catch (NoSuchAlgorithmException | InvalidAlgorithmParameterException
```
Comment (sputnikci) [inline app/src/main/java/org/mozilla/focus/activity/FingerprintActivity.java]: \[Checkstyle\] ERROR: 'catch' is not preceded with whitespace\.

Comment (sputnikci) [inline app/src/main/java/org/mozilla/focus/activity/FingerprintActivity.java]: \[Checkstyle\] ERROR: '\}' is not followed by whitespace\.

File: app/src/main/java/org/mozilla/focus/activity/FingerprintActivity.java [hunk 20]
```diff
@@ -0,0 +1,145 @@
...
+                | CertificateException | IOException e){
```
Comment (sputnikci) [inline app/src/main/java/org/mozilla/focus/activity/FingerprintActivity.java]: \[Checkstyle\] ERROR: '\{' is not preceded with whitespace\.

**Comment by pocmo** [comment]
Thank you for your code contribution!

There are a couple of reasons why I can't merge it for now:
* There are a bunch of unrelated code changes.
* Neither the filed issue nor the PR contains a description what you actually want to achieve here
* This is a rather big feature and we'd first discuss this with product and UX teams on the filed issue (That's why it's important to contain more text).
* There are a bunch of code quality issues. Without going into detail the violations sputnik commented on are a good start.

**Comment by pocmo** [comment]
We talked at our triage meeting about this and our UX/product team will have a look. I'll close this PR for now until we figure out how we want to move forward. After that we can revive this PR. Thank you again for your patch! :)

**Pull request closed without merging**